src/project.rs

```diff
@@ -0,0 +1,186 @@
+// Copyright 2024 The Fuchsia Authors
+//
+// Licensed under the 2-Clause BSD License <LICENSE-BSD or
+// https://opensource.org/license/bsd-2-clause>, Apache License, Version 2.0
+// <LICENSE-APACHE or https://www.apache.org/licenses/LICENSE-2.0>, or the MIT
+// license <LICENSE-MIT or https://opensource.org/licenses/MIT>, at your option.
+// This file may not be copied, modified, or distributed except according to
+// those terms.
+
+use crate::pointer::{
+    invariant::{Aliasing, Aligned, Alignment, AlignmentMapping, MappedAlignment, Valid},
+    Ptr,
+};
+
+// NOTE: `NAME_HASH` is not semver-stable - it just needs to be consistent
+// within a single compilation.
+//
+// TODO: `#[diagnostic::on_unimplemented]`
+#[doc(hidden)]
+pub unsafe trait Field<const NAME_HASH: u128> {
+    type Type;
+    const OFFSET: isize;
+
+    // Basically, `Preserved` unless the struct is `#[repr(packed)]`. TODO:
+    // Confirm that this is the correct behavior.
+    type AlignmentMapping: AlignmentMapping;
+}
+
+pub unsafe trait Projectable {
+    type Of<T>;
+    type Inner;
+}
+
+unsafe impl<T> Projectable for core::mem::ManuallyDrop<T> {
+    type Of<U> = core::mem::ManuallyDrop<U>;
+    type Inner = T;
+}
+
+// TODO: Use `Ptr` to make this generic over `&`, `&mut`, `Box`, etc.
+#[doc(hidden)]
+#[allow(clippy::must_use_candidate)]
+#[inline(never)]
+pub const fn project<'a, P, A, AA, const NAME_HASH: u128>(
+    // ptr: *const P,
+    ptr: Ptr<'a, P, (A, AA, Valid)>,
+) -> Ptr<
+    'a,
+    P::Of<<P::Inner as Field<{ NAME_HASH }>>::Type>,
+    (A, MappedAlignment<AA, <P::Inner as Field<{ NAME_HASH }>>::AlignmentMapping>, Valid),
+>
+where
+    P: Projectable,
+    P::Inner: Field<{ NAME_HASH }>,
+    A: Aliasing,
+    AA: Alignment,
+{
+    // NOTE: Doing the actual cast inside this function makes the `project!`
+    // macro simpler, but it also means that we can't handle both const code and
+    // unsized types at the same time. In order to be const, we can't call trait
+    // methods, but in order to support unsized types, we'd need to provide a
```

Comment:
For slice DSTs, all you need is the offset of the last field, which is the same for all sizes. This could be stored as a `const`.